rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,697 @@
+terrifying experience (restoring from a pre-upgrade snapshot is a safer
+alternative but not mentioned in the docs or logs). 
+
+Cloud users don’t have access to Kibana logs to be able to identify and remedy
+the cause of the migration failure. Apart from blindly retrying migrations by
+restoring a previous snapshot, cloud users are unable to remedy a failed
+migration and have to escalate to support which can further delay resolution. 
+
+Taken together, version upgrades are a major operational risk and discourage
+users from adopting the latest features. 
+
+# 3. Saved Object Migration Errors
+
+Any of the following classes of errors could result in a Saved Object
+migration failure which requires manual intervention to resolve: 
+
+1. A bug in a plugin’s registered document transformation function causes it
+   to throw an exception on _valid_ data.
+2. _Invalid_ data stored in Elasticsearch causes a plugin’s registered
+   document transformation function to throw an exception .
+3. Failures resulting from an unhealthy Elasticsearch cluster:
+    1. Maximum shards open
+    2. Too many scroll contexts
+    3. `circuit_breaking_exception` (insufficient heap memory)
+    4. `process_cluster_event_timeout_exception` for index-aliases, create-index, put-mappings
+    5. Read-only indices due to low disk space (hitting the flood_stage watermark)
+    6. Re-index failed: search rejected due to missing shards
+    7. `TooManyRequests` while doing a `count` of documents requiring a migration
+    8. Bulk write failed: primary shard is not active
+4. The Kibana process is killed while migrations are in progress.
+
+# 4. Design
+## 4.0 Assumptions and tradeoffs
+The proposed design makes several important assumptions and tradeoffs.
+
+**Background:**
+
+The 7.x upgrade documentation lists taking an Elacsearch snapshot as a
+required step, but we instruct users to retry migrations and perform rollbacks
+by deleting the failed `.kibana_n` index and pointing the `.kibana` alias to
+`.kibana_n-1`:
+ - [Handling errors during saved object
+migrations.](https://github.com/elastic/kibana/blob/75444a9f1879c5702f9f2b8ad4a70a3a0e75871d/docs/setup/upgrade/upgrade-migrations.asciidoc#handling-errors-during-saved-object-migrations)
+ - [Rolling back to a previous version of Kibana.](https://github.com/elastic/kibana/blob/75444a9f1879c5702f9f2b8ad4a70a3a0e75871d/docs/setup/upgrade/upgrade-migrations.asciidoc#rolling-back-to-a-previous-version-of-kib) 
+ - Server logs from failed migrations.
+
+**Assumptions and tradeoffs:**
+1. The simplicity of idempotent, coordination-free migrations outweighs the
+   restrictions this will impose on the kinds of migrations we're able to
+   support in the future. See (4.2.1)
+2. Maintaining the upgrade behaviour and rollback procedures of previous
+   Kibana 7.x releases is more important than introducing enhancements like:
+   1. (2.6) _mixed Kibana versions shouldn't cause data loss_.
```

Comment:
Is my understanding correct that the data loss occurs because after these changes all instances of Kibana will use `.kibana_${version}` for normal saved-object operations, so old versions will continue writing to the old indices after a new instance comes online?

Reply:
Although the tradeoff is probably still valid, it's misleading since the updated algorithm managed to achieve both: we can prevent data loss for mixed versions AND we still have a backup `.kibana_n` / `.kibana_${version}` index.

We won't have data-loss in your example because we set the source index to read-only. So the outdated node will probably behave in unexpected ways, but it won't acknowledge writes that don't succeed.

Reply:
Cool, even better!